lib_collections/marklogic.py

```diff
@@ -0,0 +1,243 @@
+
+    """
+    r = requests.get(
+        auth=HTTPBasicAuth(
+            MARKLOGIC_LDR_USER,
+            MARKLOGIC_LDR_PASSWORD,
+        ),
+        headers={'Content-type': 'text/turtle'},
+        params={'query': sp_query(manifid)},
+        url='http://marklogic.lib.uchicago.edu:8008/v1/graphs/sparql'
+    )
+    j = json.loads(r.content.decode('utf-8'))
+    return j
+
+
+def align_field_names(dct: dict,
+                      wagtail_field_names: list) -> dict:
+    """
+    Arranges metadata fields for object page so that they correspond to the order
+    in which those fields appear in the Wagtail admin interface.
+
+    Args: 
+        Dictionary representing result of Mark Logic query, list of field
+        names (strings) stored in Wagtail database for a given collection
+
+    Returns:
+        Dictionary representing metadata for display in object page
+    """
+    filtered_names = [item
+                      for item in wagtail_field_names
+                      if item in dct.keys()]
+    updated_dict = {key: dct[key] for key in filtered_names}
+    return updated_dict
+
+
+def triples_to_dict(dct: dict,
+                    wagtail_field_names: list):
+    """
+    Capitalizes all metadata field names and sorts them using align_field_names
+    if the Wagtail database has information about which fields to display; if
+    not, displays all of them in alphabetical order.
+
+    Args: 
+        Dictionary represengint result of Mark Logic query, list of field
+        names (strings) stored in Wagtail database for a given collection
+
+    Returns: 
+        Reformatted dictionary representing metadata
+    """
+    try:
+        results = dct['results']['bindings'][0]
+        field_dict = {
+            key.capitalize(): results[key]['value'] for key in results.keys()
+        }
+        if wagtail_field_names:
+            return align_field_names(field_dict, wagtail_field_names)
+        else:
+            return field_dict
+    except KeyError:
+        raise Exception("Mark Logic result not formatted as expected")
+    except IndexError:
+        raise Http404
+
+
+def get_record_no_parsing(manifid: str,
+                          wagtail_field_names: list) -> dict:
+    """
+    Queries Mark Logic for collection object metadata fields, leaves the values
+    of those metadata fields alone, in their raw form.
+
+    Args: 
+        NOID string, list of field names
+
+    Returns:
+        Dictionary representing metadata
+    """
+    raw_record = get_raw_record(manifid)
+    return triples_to_dict(raw_record, wagtail_field_names)
+
+
+def remove_trailing_comma(string: str) -> str:
+    """
+    Remove leading/trailing whitespace and trailing commas from a string.
+
+    Args:
+        Metadata string
+
+    Returns:
+        Cleaned up metadata string
+    """
+    stripped = string.strip()
+    rev = stripped[::-1]
+    if len(rev) > 0 and rev[0] == ',':
+        return rev[1:][::-1]
+    else:
```

Comment:
Yeah, I think we're thinking along the same lines there; it often seemed to me that the 'Args/Returns' convention was basically putting type information into the docstring.  I guess one case where it helps is when the type doesn't tell you much about what the thing actually is--for example, if it's a JSON string, the type hint will just say `str` but 'Args' could say 'JSON string'.  I experimented a little bit using a module that Python provides for fancier types (i.e. that allows for polymorphism), but I wasn't able to get it working, so I reverted to the not very interesting built-in types.  This could also be something interesting to look at going forward.